AAInstrumentation/unique_hashes.py

```diff
@@ -24,12 +25,50 @@ def get_num_unique_hashes(
         num_runs: int,
     ) -> int:
         random.seed(self.random_seed)
-        population = self.get_n_random_sequences(num_candidates, num_runs)
+        full_population = set()
         distinct_hashes = set()
+        last_len = -1
 
-        for i, sample in enumerate(population):
-            curr_hash = self.run_assemble_and_get_hash(file_name, i, sample)
-            distinct_hashes.add(curr_hash)
+        while len(full_population) < num_runs:
+            # It can be the case that some of the random sequences we generate
+            # are actually the same since changing AA queries can lead to fewer
+            # queries being present. Therefore, we loop here until we have enough
+            # sequences that are actually unique runs.
+
+            # We always at least generate 20 new sequences. This tries to make
+            # sure that we give the experiment a chance to generate new sequences
+            # if we are close to the end.
+            population = self.get_n_random_sequences(
+                num_candidates, max(num_runs - len(full_population), 20)
+            )
+
+            for i, sample in enumerate(population):
+                curr_hash = self.run_assemble_and_get_hash(file_name, i, sample)
+                new_num_candidates = self.get_candidate_count(file_name, sample)
+                full_population.add(sample)
```

Comment:
Since you already have a set, would it not make sense to check if you already have included the sample in the full_population and skip compiling etc?